### DISCUSSION TITLE
Some Tips for Using OpenH264

### DISCUSSION THREAD

**Discussion by jmdelahanty** [wehr-lab/autopilot#156] [General]
Today I spent a fair amount of time trying to get `opencv` and `openh264` from Cisco to play nice with each other. I was originally using `DIVX` encoding, but resources online and colleagues here on campus suggested I use `H264` encoding for video instead as its compression is nearly lossless and allows for reliable seeking. Although `opencv` allows you to use `H264` through it's installation of `ffmpeg`, it appears that the codec itself isn't installed and made available to *write* video to disk.

So you have to go [here](https://github.com/cisco/openh264) and build it yourself or, if you're like me and want to just use conda environments, you go [here](https://github.com/cisco/openh264/releases) and download the version your application is using. Be sure to decompress it. I couldn't get `opencv` to recognize the `.dll` that's used, however, until I placed it in the following path:
`\path\to\anaconda\env\bin\openh264-2.2.0-win64.dll` (sorry for Windows)

After that, in `opencv`, be sure to use the `avc1` `fourcc` flag in the `VideoWriter` object. From what I've seen so far, `avc1` and `H264` are essentially synonyms for the same codec.

Since other people using Autopilot might want to incorporate videos from machine vision cameras into their setups I figured I'd put this kind of info here for those that might find it helpful.

You can see more in this silly little changelog I keep [here](https://github.com/Tyelab/bruker_control/blob/main/CHANGELOG.md) if you want.  Happy piloting and Перемога України!

**Comment by sneakers-the-rat** [comment]
yes. h.264 is a codec, a specification for encoding and decoding video, and x264 is an open-source implementation. DivX is a very ye olde codec that also had an open source competitor XviD, though the relationship is a bit different. ffmpeg is a very storied body of code that collects a lot of the open source codecs of the world that is both beautiful and terrifying.

x264 and x265 can both do lossless encoding, and have a tunable quality factor -- I've spent a lot of time encouraging people to just write a quick loop and test their analyses on different compression values and see when it numerically impacts their analysis instead of spending any time reasoning about it. run a laptop overnight and empirically determine your compression.

tl;dr if you have compute and want to do archival storage use ffmpeg and x265
docs: 
- x265: https://trac.ffmpeg.org/wiki/Encode/H.265
- x264: https://trac.ffmpeg.org/wiki/Encode/H.264
- this is a class that's mostly a genicam abstraction, built around Spinnaker but shouldnt take too much to make fully general, tho does need a new driver build script bc I think it's changed a lot: https://docs.auto-pi-lot.com/en/latest/hardware/cameras.html#autopilot.hardware.cameras.Camera_Spinnaker
- thin wrapper around skvideo, which is a thin wrapper around ffmpeg, which uses optional on the fly compression to suit the needs of low memory raspi: https://docs.auto-pi-lot.com/en/latest/hardware/cameras.html#autopilot.hardware.cameras.Video_Writer

if you have docs for doing this in windows would love to have it up on the wiki or a docs pr, I am not very familiar with it!

**Reply by jmdelahanty** [reply]
I could try making a docs pr! Conda did most of the work for it, I just had to move that .dll file to the right spot so it could encode things. But I can write out what I did as a pr this week!

**Reply by jmdelahanty** [reply]
Also, can you have opencv do lossless encoding and use the tunable quality factor? I know about the `-crf` flag that you use with ffmpeg, but I couldn't seem to find how to do that with opencv.

**Reply by sneakers-the-rat** [reply]
I don't typically use opencv to encode videos if that's all I'm doing. opencv is more for realtime/framewise processing. Encoding is more of a problem of representation of a video on disk. ffmpeg gives you a hell of a lot more control over encoding. As far as I've been able to gather, opencv uses ffmpeg as a backend but only exposes the fourcc codec name [as a parameter](https://docs.opencv.org/4.5.5/dd/d9e/classcv_1_1VideoWriter.html#afec93f94dc6c0b3e28f4dd153bc5a7f0) (meanwhile, [the ffmpeg docs...](https://ffmpeg.org/ffmpeg-all.html)). 

edit: so, to be clear, i'll use one encoding strategy when acquiring (eg., writing with opencv or whatever other interface/codec that suits my constraints, like if i/o is cheap but compute is dear, i'll use a relatively fast but low compression ratio codec, and vice versa like streaming frames will compress something more before sending over the wire) and then once I've acquired it re-encode for archival purposes.

**Reply by jmdelahanty** [reply]
The main reason I landed on OpenCV was because I wasn't sure how else to take the stream of images coming off the camera and writing it to disk. If there's a better way to do it then I'd be happy to learn/switch how we write videos to disk! It looks like the CPU is never peaked by using either DIVX or H264 encoding so it doesn't seem like I have any issues with that. Do you think it's a bad move to just encode things with H264 from the get go to disk since it seems like H264 is used for archival purposes? The main thing that seems to get peaked during a recording is the bandwidth the card in the computer is capable of handling. It can only manage 500mbps and just one camera recording at 1280x1024 uses 300mbps.

Can I stream video to disk in Python without having opencv be in the middle?

**Reply by jmdelahanty** [reply]
From the opencv docs:

`If FFMPEG is enabled, using codec=0; fps=0; you can create an uncompressed (raw) video file.`

Would this be considered better practice datawise? I'm so new to encoding video streams that I honestly don't know. From what people in the lab do video is always written using an encoder through MATLAB or opencv. If there's a better way we should do it then I'm all for that!

**Reply by sneakers-the-rat** [reply]
it's almost always unnecessary to store raw video (from a camera, microscope is different matter), as compression is pretty well within the limits of most CPUs and for many video streams in science most of the image doesn't change much so even slight compression can reduce the size by hundreds of times. If I/O is the limiting point, then ya definitely want to encode in some way before writing, as 1280x1024 encoded as a raw avi is, as you said, 300MB/s (a 10s clip is 3.16GB). (this is a pet peeve of mine because scientists are always like ~ the data from my experiments is soooo big and so that's why i can't share it and need a ton of AWS space to store it ~ when in reality if they compressed it it would go from 1TB/experiment to like 500MB/experiment. Literally that same 3.16GB raw video -> 1.5MB with x265, or 2100x smaller, with no detectable difference).

Basically during acquisition you want to store something in whatever format you can that has the quality/etc. you need, and then what I do is write something to automatically go through and clean up all my data into whatever form i want to archive it in. So in this case, by all means write using opencv during acquisition (or use the ffmpeg interfaces chris and i have mentioned), and as long as that's what you need, you can always re-encode later. 

The only downside I can think of using opencv + ffmpeg to encode h.264 is that just from the fourcc code it probably uses the default settings of the codec, which is `-crf 23`, so if that's causing problems for you then i would acquire it raw and then encode afterwards. as the linked documentation above describe, x264 has a lossless mode if it's needed (it probably won't be)

(there is never a time when you would want to use divx, it's a very old codec, worse compression, more artifacts, larger size).

**Reply by jmdelahanty** [reply]
Thanks for the tips Jonny! I'll definitely stay using the H264 codec during the experiment then. As far as I'm aware, there hasn't been a problem with using the standard settings of the codec for either DIVX or H264 and it's high enough quality video for us to get what we need from it. There's no visible difference between an original video that's been encoded with DIVX and one that's been re-encoded with H264, both still look pretty nice.

**Reply by sneakers-the-rat** [reply]
to be clear: DIVX is a very low quality codec that is outdated along every axis, so the comparison is between uncompressed/lossless video and compressed video, of which h.264 is one method, rather than between divx and h.264. if both are good enough then that's great, but just so the concept is clear.

**Reply by jmdelahanty** [reply]
I see, thank you for the clarification! I had no idea DIVX was that old and low quality. I'm glad to be past it now lol.

**Comment by cxrodgers** [comment]
I would definitely use ffmpeg for compressing video. It is really good at
that one thing. Things like Matlab don’t compress video well in my
experience and it sounds like opencv is just doing it’s compression via a
wrapper around ffmpeg anyway. It’s annoying to figure out the syntax the
first time but after that it should just work.

It’s best to get an up to date version of ffmpeg directly from their
website. There are lots of old versions floating around (eg in linux
repositories) and if you have problems their community will say it’s
because you’re not using the most recent version.

You can also acquire video directly with ffmpeg from the camera and encode
in one step, depending on what camera you use. I’ve done this with a webcam
that uses the v4l2 standard. If that’s not possible with your camera, I
would acquire in as raw a format as possible, compress with ffmpeg (using a
compression level that doesn’t visibly change the video quality), and
discard the raw when needed to save space.

good luck!

**Reply by jmdelahanty** [reply]
Hello! Thank you for the tips! I'm thankfully getting through using ffmpeg for our videos so far in opencv. The camera we have is a Teledyne Genie Nano for this application and I don't think it does any compression/encoding on it's own. I use a package called [Harvesters](https://github.com/genicam/harvesters) to grab the data coming of the camera which represents each frame as a numpy array. I then pass the content into the opencv writer using the h264 codec. I haven't been able to see if I can add additional parameters to the writer though. As Jonny linked above, there's so many additional things ffmpeg can do that I'm not sure opencv knows how to pass along. Any clues about how I might pass the arrays into ffmpeg directly so I don't have to rely on opencv to do it for me?

**Reply by cxrodgers** [reply]
If you already have images as numpy arrays, you can directly write to an ffmpeg process which will compress the video (or store losslessly, depending on the parameters you provide). Check out my answer on stack overflow: https://stackoverflow.com/questions/4092927/generating-movie-from-python-without-saving-individual-frames-to-files/29525514#29525514
Since this needs to run in real-time without dropping frames, you probably don't want a really heavy compression, which will eat up precious CPU cycles.

Another option is using the python module `ffmpeg-python` https://github.com/kkroening/ffmpeg-python. (Note: NOT the same as python-ffmpeg!) You can script complex chains of ffmpeg processes. This is essentially similar to my approach, but there's a whole package for it.

Finally, this probably won't work for you since I guess you need Harvesters to capture video, but I have used something like this to capture video from /dev/video0, compress it, and tee the output to a file and also to a separate ffplay process displaying the video:

`ffmpeg -f video4linux2 -i /dev/video0 -vcodec libx264 -qp 0 -vf format=gray -preset ultrafast -f rawvideo - | tee data.mp4 | ffplay -analyzeduration 500000 -`

The common element in all these solutions is the `-f rawvideo` flag, which allows you to directly provide binary images to ffmpeg.

This could get complicated! General advice still stands: If you have something working already, always consider leaving well enough alone, and optimizing only when need be.

**Reply by sneakers-the-rat** [reply]
(agreed with all this^^ for the record, `skvideo` also basically does what Chris described in that stackexchange post, and so that's also what the `autopilot.hardware.cameras.Video_Writer` class does) 

https://github.com/scikit-video/scikit-video/blob/87c7113a84b50679d9853ba81ba34b557f516b05/skvideo/io/ffmpeg.py#L93-L108

and

https://github.com/scikit-video/scikit-video/blob/87c7113a84b50679d9853ba81ba34b557f516b05/skvideo/io/abstract.py#L512-L515

**Reply by jmdelahanty** [reply]
Thank you for the posts! This is great information. I'll have to toy around with using `skvideo`, I haven't tried it all before!

**Reply by jmdelahanty** [reply]
Hey again everyone! So I'm messing with `skvideo` now and am wondering how to change a part of how things are encoded to disk. Currently, it appears that our videos are written as having a color channel. The issue is that we don't have a color channel to record from! The camera, as far as I'm aware, only produces grayscale images since it's a monochrome camera. I've tried using `np.squeeze` on frames from my video, but that doesn't seem to do anything.

So I'm wondering if I can use `skvideo` to specify that I want data written to disk in grayscale. I perused the `ffmpeg` docs but didn't see a clear answer. Same with `skvideo`.

Any advice?

**Comment by cxrodgers** [comment]
I used to use "pix_fmt gray" to tell ffmpeg to encode in grayscale. But it
caused all kinds of compatibility problems with other things trying to read
the video (I don't remember what). I think it doesn't matter, because it
will encode gray as equal values of R, G, and B, which will be extremely
easy for any compression algorithm (such as libx264) to compress by 3x. So
if your concern is disk space, I think this is not an issue.

**Reply by jmdelahanty** [reply]
Thank you so much for the tip!

The main reason I was thinking of trying to do that was simply to make the data I was working with easier to process/faster to process later. [Here's](https://dask.discourse.group/t/performing-hog-matrices-on-pims-chunks-through-imageio/570) why I was looking into it. We're calculating HOGs from videos and I'm trying to use `pims` and `dask` to make cropping and HOG calculations fast as well as make it so we can just have one datastream instead of generating different video clips. It seems like having the color channel slows things down a bit and also makes the resulting dataset quite a lot larger. The original repository we're going off of ran things over individual images (so thousands of .jpg files) but with `pims` we can do it over videos! I thought it would be easier but I'm pretty stuck lol

If you have any advice on how to solve the problems I'm running into there feel free to comment!

**Comment by cxrodgers** [comment]
I ran into the same problem with the feature detector network in
DeepLabCut. It was originally designed for color videos (humans) but most
animal researchers have grayscale video. It's wasteful to dump 3x as many
datapoints into the network. But, it was too hard to fix, so I just
accepted it. I think you will find that a lot, most video processing
pipelines are really designed around color, and if you really want
grayscale you will have to re-engineer things.

If you literally just want to convert color frames to grayscale, you can
use
https://scikit-image.org/docs/dev/auto_examples/color_exposure/plot_rgb_to_gray.html
You basically just have to mean the (identical) RGB channels to get a
luminance channel. np.squeeze does not work because the extra dimension is
not empty, but rather contains identical data in RGB. But if your
downstream application wants color, you will have to give it color.

**Reply by sneakers-the-rat** [reply]
+1 this. no video compression algorithm (storage format) is really designed for grayscale, so just lose two of the channels when you load it (decompress) and then duplicate the channels again when you go back to storage (compress). Any time you are processing video data (except if you are positive it will always be very small/short videos) you should do it either iteratively or in chunks rather than trying to load the entire thing into memory.

**Comment by jmdelahanty** [comment]
An update to this just in case people run into this:

If you use miniconda (I screwed up my anaconda distro so I nuked it and started over...) the path for where this `.dll` goes isn't in `bin` anymore but just in the env folder itself. My miniconda environment for the project doesn't have a bin folder for some reason...

**Comment by jmdelahanty** [comment]
An update to this adventure I've been taking! I just got skvideo to work for recording videos at the same time as our microscope recording using skvideo! It was so much easier than I thought it would be! The other exciting piece is that with the help from you @cxrodgers about multiprocessing I got that to run in my script too without blowing things up! I was very excited so I wanted to post here about it.
